### ISSUE TITLE
Reactivate fast squaring algorithms

### ISSUE THREAD

**Issue by mratsim** [mratsim/constantine#68]
Commit https://github.com/mratsim/constantine/pull/58/commits/297196586221a4b8c8775dd519e7695e33c8da2a
deactivates the fast path for generic squaring as there is an off-by-one on 32-bit with inputs from the test suite (from #61 #62):

https://github.com/mratsim/constantine/blob/297196586221a4b8c8775dd519e7695e33c8da2a/tests/t_finite_fields_mulsquare.nim#L163-L214

To be reactivated:
https://github.com/mratsim/constantine/blob/297196586221a4b8c8775dd519e7695e33c8da2a/constantine/arithmetic/limbs_montgomery.nim#L246-L316

**Comment by mratsim** [comment]
In the Montgomery domain, the off-by-one is the least significant bit

![image](https://user-images.githubusercontent.com/22738317/85337945-bffea780-b4e1-11ea-9680-1e01bee8b383.png)

Also it doesn't happen only for generic square but also the fast square `montySquare_CIOS_nocarry` from Consensys article: https://hackmd.io/@zkteam/modular_multiplication which should be straightforward to implement

```
 for i=0 to N-1
    C, t[i] = a[i] * a[i] + t[i]
    p = 0
    for j=i+1 to N-1
        p,C,t[j] = 2*a[j]*a[i] + t[j] + (p,C)

    A = C

    m = t[0] * q'[0]
    C, _ = t[0] + q[0]*m
    for j=1 to N-1
        C, t[j-1] = q[j]*m +  t[j] + C

    t[N-1] = C + A
```

vs 

 https://github.com/mratsim/constantine/blob/0400187f05d9b9cd3198b68eb72780cb2e983e83/constantine/arithmetic/limbs_montgomery.nim#L208-L244

**Comment by mratsim** [comment]
unfused squaring was added in https://github.com/mratsim/constantine/pull/144
including assembly version though no MULX/ADOX/ADCX implemented as it's quite tricky in terms of registers (commented out).

The normal assembly version is as fast as multiplication with MULX/ADOX/ADCX.

**Comment by mratsim** [comment]
Closed in #160
